capsules/src/segger_rtt.rs

```diff
@@ -91,77 +91,79 @@
 //! ```
 
 use core::cell::Cell;
-use kernel::common::cells::{OptionalCell, TakeCell};
+use core::marker::PhantomData;
+use kernel::common::cells::{OptionalCell, TakeCell, VolatileCell};
 use kernel::hil;
 use kernel::hil::time::Frequency;
 use kernel::hil::uart;
 use kernel::ReturnCode;
 
-/// Buffer for transmitting to the host.
-pub static mut UP_BUFFER: [u8; 1024] = [0; 1024];
-
-/// Buffer for receiving messages from the host.
-pub static mut DOWN_BUFFER: [u8; 32] = [0; 32];
```

Comment:
Can we leave the numbers here? This is a hint from the capsule author to everyone else on what size these buffers should be.

Reply:
I disagree that this hint makes sense in the capsule. The capsule's logic doesn't depend on these sizes at all, and one may want to adapt the sizes in various scenarios:

- On a board with limited memory, reducing the sizes of the buffers in all included capsules allows to increase the memory available for other parts of the system (e.g. adding more capsules, giving more memory to userspace).
- On the contrary, if one wants to increase the size of these buffers to e.g. increase debugging throughput, this flexibility should be easily available at the board or component level, without having to dig into the capsule and break all the other boards.

The rationale applies as in https://github.com/tock/tock/pull/1547.

Besides, `static mut` is problematic as highly `unsafe` in itself (see again https://github.com/tock/tock/issues/1545, https://github.com/rust-lang/rust/issues/53639 and https://users.rust-lang.org/t/why-is-static-mut-bad/32888/). Given that capsules are `forbid(unsafe)`, I find it weird to encourage putting `static mut` in capsules. On the other hand, centralizing all static initialization with `static_init` at the boards & components level makes it easier to reason about.

Reply:
Using `static_init!()` is fine, I just think `const DEFAULT_UP_BUFFER_LENGTH: usize = 1024` would be a useful hint to board authors who don't then have to guess what a "good" length would be. Certain sensors in particular have a minimum length for correctness, and I would imagine that performance likely depends on the sizing.

Boards could then use their own sizes, or use the default value.